tool/tsh/db.go

```diff
@@ -351,6 +381,113 @@ func onDatabaseConnect(cf *CLIConf) error {
+	database, err := pickActiveDatabase(cf)
+	if err == nil {
+		return database, nil
+	}
+	if !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	db, err := getDatabaseResourceByName(cf, tc, dbName)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &tlsca.RouteToDatabase{
+		ServiceName: db.GetName(),
+		Protocol:    db.GetProtocol(),
+		Username:    cf.Username,
+		Database:    cf.DatabaseName,
+	}, nil
+}
+
+func getDatabaseResourceByName(cf *CLIConf, tc *client.TeleportClient, dbName string) (types.Database, error) {
+	var databases []types.Database
+	err := client.RetryWithRelogin(cf.Context, tc, func() error {
+		allDatabases, err := tc.ListDatabases(cf.Context)
+		for _, database := range allDatabases {
+			if database.GetName() == dbName {
+				databases = append(databases, database)
+			}
+		}
+		return trace.Wrap(err)
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if len(databases) == 0 {
+		return nil, trace.NotFound(
+			"database %q not found, use 'tsh db ls' to see registered databases", dbName)
+	}
+	return databases[0], nil
+}
+
+func dbCertExpiredOrMFAIsRequired(cf *CLIConf, tc *client.TeleportClient, database *tlsca.RouteToDatabase, profile *client.ProfileStatus) (bool, error) {
+	// Load DB cert and check if cert is still valid. Missing cert is treated as expired cert.
+	certExpired, err := isExpiredCert(profile.DatabaseCertPath(database.ServiceName))
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return true, nil
+		}
+		return false, trace.Wrap(err)
+	}
+	if certExpired {
+		return true, nil
+	}
+	// Call API and check is a user needs to use MFA to connect to the database.
+	mfaRequired, err := isMFADatabaseAccessRequired(cf, tc, database)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	return mfaRequired, nil
+}
+
+// isMFADatabaseAccessRequired calls the IsMFARequired endpoint in order to get from user roles if access to the database
+// requires MFA.
+func isMFADatabaseAccessRequired(cf *CLIConf, tc *client.TeleportClient, database *tlsca.RouteToDatabase) (bool, error) {
+	proxy, err := tc.ConnectToProxy(cf.Context)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	cluster, err := proxy.ConnectToCluster(cf.Context, tc.SiteName, true)
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	defer cluster.Close()
+
+	dbParam := proto.RouteToDatabase{
+		ServiceName: database.ServiceName,
+		Protocol:    database.Protocol,
+		Username:    cf.Username,
+		Database:    database.Database,
+	}
+	mfaResp, err := cluster.IsMFARequired(cf.Context, &proto.IsMFARequiredRequest{
+		Target: &proto.IsMFARequiredRequest_Database{
+			Database: &dbParam,
+		},
+	})
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+	return mfaResp.GetRequired(), nil
+}
+
+func databaseLoginOnMFAOrExpiredCert(cf *CLIConf, tc *client.TeleportClient, database *tlsca.RouteToDatabase, profile *client.ProfileStatus) error {
```

Comment:
I'm generally not a fan of such verbose function names, feels very Java-esque :) The way I would do it is just have a `needRelogin()` function that onDatabaseConnect calls and then can call databaseLogin depending on the result (basically move logic to the caller).